components/GAProvider.tsx

```diff
@@ -0,0 +1,106 @@
+'use client';
+import { useEffect } from 'react';
+import { setCookie, getCookie } from 'cookies-next';
+import { get } from 'http';
+
+export default function GAProvider({ children }: { children: React.ReactNode }) {
+  const cookieName = 'SU-GC-UTMs';
+
+  useEffect(() => {
+    if (typeof window === 'undefined') {
+      return;
+    }
+
+    const urlParams = new URLSearchParams(window.location.search);
+    const utm_source = urlParams.get('utm_source') || undefined;
+    const utm_medium = urlParams.get('utm_medium') || undefined;
+    const utm_campaign = urlParams.get('utm_campaign') || undefined;
+    const utm_content = urlParams.get('utm_content') || undefined;
+    const utm_term = urlParams.get('utm_term') || undefined;
+    const UTMs = {
+      utm_source,
+      utm_medium,
+      utm_campaign,
+      utm_content,
+      utm_term,
+    };
+
+    // UTM source is a required param for GA to work.
+    if (utm_source) {
+      setCookie(cookieName, UTMs, {
+        path: '/',
+        domain: window.location.hostname,
+        secure: window.location.protocol === 'https' ? true : false,
+        httpOnly: false,
+        sameSite: 'strict',
+      });
+    }
+
+    // Function to get the active anchor element that was clicked on if there is one.
+    const getActiveElement = (event: MouseEvent):Element | null => {
+
+      const target = event.target;
+      if (target instanceof Element === false) { return null; }
+
+      // If the target is a link.
+      if ((target as Element).tagName === 'A') { return target as Element; }
+
+      // Start with the parent element of the target
+      let parentElement = (target as Element).parentElement;
+
+      // Loop through the parent elements until an anchor element is found or parent becomes null
+      while (parentElement !== null) {
+        // Check if the current parent element is an anchor element
+        if (parentElement.tagName === 'A') {
+          return parentElement; // Found an anchor element, so return it
+        }
+
+        parentElement = parentElement.parentElement; // Move up to the next parent element
+      }
```

Comment:
Might be able to use something like `target.closest('a')` if you're looking to find the nearest link ancestor (including self). Save a few lines of code.

https://developer.mozilla.org/en-US/docs/Web/API/Element/closest